In [2]:
import pandas as pd
import numpy as np

In [4]:
df = pd.read_csv('/content/Advanced_pandas_sales_dataset.csv')
df

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
...,...,...,...,...,...,...,...,...,...,...,...
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00
226,1227,2024-04-02,Dev,Pune,South,Mouse,Accessories,1669,6,0.05,323.13
227,1228,2024-01-14,Vikas,Chennai,North,Tablet,Electronics,2682,7,0.15,119.82
228,1229,2024-09-04,Priya,Pune,East,Tablet,Electronics,2635,8,0.00,531.76


## **Q1 Dataset Exploration**

**Perform the following exploratory steps:**

- Display the first 10 rows

- Display the dataset structure using info()

- Find the number of rows and columns

- Identify the data types of each column

In [5]:
# Ans
# Display the first 10 rows

df.head(10)


,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
6,1007,2024-06-24,Dev,Chennai,East,Monitor,Electronics,1170,5,0.20,22.83
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
9,1010,2024-12-25,Pooja,Ahmedabad,North,Phone,Electronics,2711,8,0.00,477.32


In [6]:
# Display the dataset structure using info()

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230 entries, 0 to 229
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Order_ID    230 non-null    int64  
 1   Order_Date  230 non-null    object 
 2   Customer    230 non-null    object 
 3   City        230 non-null    object 
 4   Region      230 non-null    object 
 5   Product     230 non-null    object 
 6   Category    230 non-null    object 
 7   Sales       230 non-null    int64  
 8   Quantity    230 non-null    int64  
 9   Discount    230 non-null    float64
 10  Profit      230 non-null    float64
dtypes: float64(2), int64(3), object(6)
memory usage: 19.9+ KB


In [7]:
# Find the number of rows and columns

df.shape

(230, 11)

In [8]:
# Identify the data types of each column

df.dtypes

,0
Order_ID,int64
Order_Date,object
Customer,object
City,object
Region,object
Product,object
Category,object
Sales,int64
Quantity,int64
Discount,float64


## **Q2 Sales Filtering Analysis**

**Filter the dataset to display:**

- Orders where Sales > 3500

- Orders where Region = "East" and Sales > 2500

- Explain what insights you observe.

In [10]:
# Ans 2
# Orders where Sales > 3500

df[df['Sales'] > 3500]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
10,1011,2024-06-30,Neha,Delhi,South,Printer,Office,5977,6,0.00,1578.32
...,...,...,...,...,...,...,...,...,...,...,...
219,1220,2024-03-22,Meera,Bangalore,West,Phone,Electronics,5986,7,0.05,986.70
222,1223,2024-03-04,Meera,Ahmedabad,South,Phone,Electronics,4356,6,0.10,416.87
223,1224,2024-01-07,Dev,Bangalore,East,Tablet,Electronics,4115,4,0.05,870.03
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00


In [23]:
# Orders where Region = "East" and Sales > 2500

df[(df['Region'] == 'East') & (df['Sales'] > 2500 )]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
13,1014,2024-03-18,Anita,Bangalore,East,Mouse,Accessories,3616,4,0.05,463.93
21,1022,2024-01-14,Nitin,Chennai,East,Monitor,Electronics,4827,6,0.05,912.34
27,1028,2024-07-21,Dev,Kolkata,East,Tablet,Electronics,4054,6,0.00,1013.83
33,1034,2024-12-15,Nitin,Ahmedabad,East,Tablet,Electronics,4378,2,0.20,-104.30
34,1035,2024-02-20,Nitin,Ahmedabad,East,Laptop,Electronics,2535,4,0.10,188.15
41,1042,2024-07-15,Rohan,Chennai,East,Laptop,Electronics,3268,8,0.15,295.15
43,1044,2024-09-20,Kavya,Hyderabad,East,Printer,Office,2618,7,0.05,454.64


# **what insights you observe ?**

- I understand that if we compare the entire data with how many people who are spending more than 3500, we can see there are a significant number of people who can spend that much amout of money  or even more than that. There are 104 out of total 230 orders where the requirement satisfied.

- If we compare Region out of 230 orders there are 72 orders are coming form East region and if we want to apply sales where amount is more than 2500, then there are 48 such orders out of 72. That means East region is the prominant region out of four.

## **Q3 Customer Purchase Analysis**

Identify the top 5 customers based on total sales.

Hint: Use GroupBy with aggregation.

In [29]:
# Ans 3
# fist find which customer contribute how much in total sales.

customer_sales = df.groupby('Customer')['Sales'].sum()
customer_sales

,Sales
Customer,
Amit,34608
Anita,40601
Arjun,36091
Dev,48500
Isha,46818
Karan,40162
Kavya,36146
Meera,29762
Neha,41955


In [30]:
# Now find top 5 customer contributing most in sales

top_5_customers = customer_sales.nlargest(5)
top_5_customers

,Sales
Customer,
Nitin,59820
Raj,58416
Rahul,52291
Dev,48500
Isha,46818


## **Q4 Regional Sales Performance**

Calculate the total sales generated by each region.

Then identify:

Which region generated the highest revenue?

In [32]:
# Ans 4
# Total sales generated by each region.

region_sales = df.groupby('Region')['Sales'].sum()
region_sales

,Sales
Region,
East,232675
North,158210
South,183108
West,170227


In [36]:
# Region generated the highest revenue

highest_revenue_region = region_sales.idxmax()
highest_revenue_region

'East'

## **Q5 Product Performance Analysis**

Calculate the total quantity sold for each product.

Which product appears to be the most popular?



In [38]:
# Ans 5
# The total quantity sold for each product

quantity_sold = df.groupby('Product')['Quantity'].sum()
quantity_sold

,Quantity
Product,
Headphones,115
Keyboard,110
Laptop,140
Monitor,156
Mouse,122
Phone,151
Printer,108
Tablet,150


In [39]:
# Which product appears to be the most popular?

most_popular_product = quantity_sold.idxmax()
most_popular_product

'Monitor'

## **Q6 Multi-Level Grouping**

Group the dataset by: Region and Category

Calculate the total sales for each combination.

What insights can you derive from this analysis?

In [43]:
# Ans 6
# Group the dataset by: Region and Category and Calculate the total sales for each combination.

region_category_sales = df.groupby(['Region', 'Category'])['Sales'].sum().reset_index()
region_category_sales

,Region,Category,Sales
0,East,Accessories,57646
1,East,Electronics,160907
2,East,Office,14122
3,North,Accessories,62052
4,North,Electronics,70023
5,North,Office,26135
6,South,Accessories,80914
7,South,Electronics,86690
8,South,Office,15504
9,West,Accessories,64811


## **What insights can derive from this analysis?**

If we look closely then we can find that among all the four regions "Electronics" sold more then comes "Accessories" following with "Office". That means "Electonics" is the category which contribute more.

## **Q7 Pivot Table Creation**

Create a pivot table showing:

- Rows → Region

- Columns → Product

- Values → Total Sales

Which region sells the most laptops?

In [47]:
# Ans 7
# Create a pivot table

pd.pivot_table(df, index = 'Region', columns= 'Product' ,values = 'Sales',aggfunc= 'sum' )

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,23899,10845,55568,31299,22902,39667,14122,34373
North,15450,20538,4540,14922,26064,30535,26135,20026
South,24111,23480,23029,19167,33323,18218,15504,26276
West,17159,38499,10826,18178,9153,27850,21288,27274


**Which region sells the most laptops?**

East region sold most number of laptops



## **Q8 Profitability Analysis**

Calculate the average profit for each product category.

Which category is most profitable?

In [49]:
# Ans 8
# Calculate the average profit for each product category.

avg_profit = df.groupby('Category')['Profit'].mean()
avg_profit

,Profit
Category,
Accessories,410.603086
Electronics,396.462114
Office,471.938462


In [50]:
# Which category is most profitable?

avg_profit.idxmax()


'Office'

## **Q9 Discount Impact Study**

**Analyze the relationship between discount and profit.**

Tasks:

Calculate the average profit for each discount level

Identify whether higher discounts lead to lower profit



In [52]:
# Ans 9
# Calculate the average profit for each discount level

avg_profit_discount = df.groupby('Discount')['Profit'].mean()
avg_profit_discount

,Profit
Discount,
0.00,751.210588
0.05,531.368039
0.10,346.460513
0.15,237.191400
0.20,90.026923


In [64]:
# Identify whether higher discounts lead to lower profit

print(avg_profit_discount.idxmin(),'-', avg_profit_discount.min())


0.2 - 90.02692307692308


As per the above analysis we can see that when we are increasing discounts our profits falls significantly in each of the discount levels.

## **Q10 End-to-End Data Wrangling Task**

**Perform the following workflow:**

- Filter the dataset where Sales > 3000

- Group data by Region and Product

- Calculate total sales

- Reshape the result into a pivot table

Explain the business insights obtained from this analysis.

In [67]:
# Ans 10
# Filter the dataset where Sales > 3000

df2 = df[df['Sales']> 3000]
df2

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
...,...,...,...,...,...,...,...,...,...,...,...
219,1220,2024-03-22,Meera,Bangalore,West,Phone,Electronics,5986,7,0.05,986.70
222,1223,2024-03-04,Meera,Ahmedabad,South,Phone,Electronics,4356,6,0.10,416.87
223,1224,2024-01-07,Dev,Bangalore,East,Tablet,Electronics,4115,4,0.05,870.03
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00


In [77]:
# Group data by Region and Product

df2.groupby(['Region','Product'])['Sales'].sum().reset_index()

,Region,Product,Sales
0,East,Headphones,14469
1,East,Keyboard,9642
2,East,Laptop,44096
3,East,Monitor,16420
4,East,Mouse,19000
5,East,Phone,29500
6,East,Printer,3696
7,East,Tablet,29229
8,North,Headphones,5326
9,North,Keyboard,15653


In [75]:
# Calculate total sales

df2.groupby('Region')['Sales'].sum()

,Sales
Region,
East,166052
North,98062
South,141048
West,116551


In [78]:
# Reshape the result into a pivot table

df3 = pd.pivot_table(df2, index= 'Region', columns= 'Product', values= 'Sales', aggfunc= 'sum')
df3

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,14469,9642,44096,16420,19000,29500,3696,29229
North,5326,15653,4540,5087,22209,24670,17185,3392
South,10026,19249,17950,14921,26928,18218,14143,19613
West,13898,33380,5922,14354,3511,15653,12648,17185


## **The business insights obtained from this analysis.-**

- If we can see the above table carefully, we can understand that in "East" region "Laptop" sales is much more higher than any other product catergory in that same region and it is also the higher in comparison with any other region as well, even it's sales number is much higher than any other product or Laptop sales in any other region, Which gives an insight that laptop is most selling product in the company.

- Another insight which we can draw that "Laptop" sales in "North" and "West" region is significantly low in comparison with other two regions. Compay should focus on improving the sales of "Laptop" in this regions as well because having only one prominant region with most of the sales will harm the company if sales number drops in this regions.

- Also we can see that "South" is the region where sales numbers are quite well distributed among all the products. Company should apply this region's sales tactics to fix sales in other regions.

- This table also provide info about how much inventory is required to be availabe in which region according to their most selling product, so that sales would not impact for inventoty shortage.